modules/library/metadata/src/main/java/org/geotools/factory/CategoryRegistry.java

```diff
@@ -0,0 +1,200 @@
+package org.geotools.factory;
+
+import org.geotools.util.PartiallyOrderedSet;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static org.geotools.util.Utilities.stream;
+
+/*
+ * TODO: document all
+ *
+ *  - no null checks
+ *  - no subclass relationships between categories and instances
+ */
+
+class CategoryRegistry {
+
+	private final FactoryRegistry factoryRegistry;
+	private final Map<Class<?>, InstanceRegistry<?>> categories = new HashMap<>();
+
+	public CategoryRegistry(FactoryRegistry factoryRegistry) {
+		this.factoryRegistry = factoryRegistry;
+	}
+
+	public void registerCategory(final Class<?> category) {
+		// TODO: should anything happen if the category is already registered?
+		categories.put(category, new InstanceRegistry<>(factoryRegistry, category));
```

Comment:
So registering the same category multiple times should not lead to an error, right?